crates/containerd-shim-wasmtime/src/instance.rs

```diff
@@ -3,74 +3,172 @@ use containerd_shim_wasm::container::{
-
-        stdio.redirect()?;
-
-        log::info!("building wasi context");
-        let wctx = wasi_builder.build();
-
-        log::info!("wasi context ready");
         let WasiEntrypoint { path, func } = ctx.wasi_entrypoint();
 
-        let module = match ctx.wasm_layers() {
+        let data = match ctx.wasm_layers() {
             [] => {
                 log::info!("loading module from file");
                 let path = path
                     .resolve_in_path_or_cwd()
                     .next()
                     .context("module not found")?;
-
-                Module::from_file(&self.engine, path)?
+                std::fs::read(path)?
             }
             [module] => {
                 log::info!("loading module wasm OCI layers");
-                Module::from_binary(&self.engine, &module.layer)?
+                module.layer.clone()
             }
             [..] => bail!("only a single module is supported when using images with OCI layers"),
         };
 
-        let mut linker = Linker::new(&self.engine);
+        let status = match wasm_binary_type(&data) {
+            Some(WasmBinaryType::Module) => {
+                let module = Module::from_binary(&self.engine, &data)?;
+                let mut wasi_builder = WasiCtxBuilder::new();
+                wasi_builder
+                    .args(ctx.args())?
+                    .envs(envs.as_slice())?
+                    .inherit_stdio()
+                    .preopened_dir(root_path, "/")?;
+
+                stdio.redirect()?;
+
+                log::info!("building wasi context");
+                let wctx = wasi_builder.build();
 
-        wasmtime_wasi::add_to_linker(&mut linker, |s| s)?;
-        let mut store = Store::new(&self.engine, wctx);
+                let mut module_linker = wasmtime::Linker::new(&self.engine);
 
-        log::info!("instantiating instance");
-        let instance: wasmtime::Instance = linker.instantiate(&mut store, &module)?;
+                wasmtime_wasi::add_to_linker(&mut module_linker, |s| s)?;
+                let mut store = Store::new(&self.engine, wctx);
+
+                log::info!("instantiating instance");
+                let instance: wasmtime::Instance =
+                    module_linker.instantiate(&mut store, &module)?;
+
+                log::info!("getting start function");
+                let start_func = instance
+                    .get_func(&mut store, &func)
+                    .context("module does not have a WASI start function")?;
+
+                log::debug!("running {path:?} with start function {func:?}");
+
+                let status = start_func.call(&mut store, &[], &mut []);
+                status
+            }
+            Some(WasmBinaryType::Component) => {
+                let component = Component::from_binary(&self.engine, &data)?;
+                let file_perms = wasi_preview2::FilePerms::all();
+                let dir_perms = wasi_preview2::DirPerms::all();
```

Comment:
I believe this is the way it's done today but in the case of the OCI layers I don't think we need these permissions.  Any thoughts on how we can make this more configurable?  It feels like this is where a OCI artifact with configuration that had info like this in it would come into play

Reply:
I guess we need to figure out the story for this.
I can see the argument where te runtime provides an fs component with access to the container root, but you can choose to use your own fs component instead, and ignore/proxy the one provided by the runtime.

If we want to introduce config files as custom layers, it would be good to have a discussion with a wider group (wasi-wg, or even wasi-cg)

Reply:
> If we want to introduce config files as custom layers, it would be good to have a discussion with a wider group (wasi-wg, or even wasi-cg)

They may have already thought of possible solutions, or many people even outside container (e.g., wasmtime itself) could have a similar pain point.

Reply:
> If we want to introduce config files as custom layers, it would be good to have a discussion with a wider group (wasi-wg, or even wasi-cg)

+1 to wider discussion, wasn't intending to figure it out on this PR just calling it out.   I don't think it should be a custom layer either, that feels like its too open. I think this is more evidence that we really want a "real" "OCI artifact" with its own custom config media type that contains WASI specific meta data

Reply:
Raised an issue here for further discussion: https://github.com/containerd/runwasi/issues/413